
import os
import sys
import argparse
from multiprocessing import Pool
import multiprocessing as multi
from mutmap.utils import time_stamp
from mutmap.__init__ import __version__


class Params(object):

    def __init__(self, program_name):
        self.program_name = program_name

    def set_options(self):
        if self.program_name == 'mutmap':
            parser = self.mutmap_options()
        elif self.program_name == 'mutplot':
            parser = self.mutplot_options()

        if len(sys.argv) == 1:
            args = parser.parse_args(['-h'])
        else:
            args = parser.parse_args()
        return args

    def mutmap_options(self):
        parser = argparse.ArgumentParser(description='MutMap version {}'.format(__version__),
                                         formatter_class=argparse.RawTextHelpFormatter)
        parser.usage = ('mutmap -r <FASTA> -c <BAM|FASTQ> -b <BAM|FASTQ>\n'
                        '              -n <INT> -o <OUT_DIR> [-T] [-e <DATABASE>]')

        # set options
        parser.add_argument('-r',
                            '--ref',
                            action='store',
                            required=True,
                            type=str,
                            help='Reference fasta.',
                            metavar='')

        parser.add_argument('-c',
                            '--cultivar',
                            action='append',
                            required=True,
                            type=str,
                            help=('fastq or bam of cultivar. If you specify\n'
                                  'fastq, please separate pairs by commma,\n'
                                  'e.g. -c fastq1,fastq2. You can use this\n'
                                  'optiion multiple times'),
                            metavar='')

        parser.add_argument('-b',
                            '--bulk',
                            action='append',
                            required=True,
                            type=str,
                            help=('fastq or bam of mutnat bulk. If you specify\n'
                                  'fastq, please separate pairs by commma,\n'
                                  'e.g. -b fastq1,fastq2. You can use this\n'
                                  'optiion multiple times'),
                            metavar='')

        parser.add_argument('-n',
                            '--N-bulk',
                            action='store',
                            required=True,
                            type=int,
                            help='Number of individuals in mutant bulk.',
                            metavar='')

        parser.add_argument('-o',
                            '--out',
                            action='store',
                            required=True,
                            type=str,
                            help=('Output directory. Specified name must not\n'
                                  'exist.'),
                            metavar='')

        parser.add_argument('-t',
                            '--threads',
                            action='store',
                            default=2,
                            type=int,
                            help=('Number of threads. If you specify the number\n'
                                  'below one, then MutMap will use the threads\n'
                                  'as many as possible. [2]'),
                            metavar='')

        parser.add_argument('-T',
                            '--trim',
                            action='store_true',
                            default=False,
                            help='Trim fastq by trimmomatic.')

        parser.add_argument('--trim-params',
                            action='store',
                            default='33,TruSeq3-PE.fa:2:30:10,20,20,4:15,75',
                            type=str,
                            help=('Parameters for trimmomatic. Input parameters\n'
                                  'must be separated by comma with following\n'
                                  'order: phred,ILLUMINACLIP,LEADING,TRAILING,\n'
                                  'SLIDINGWINDOW,MINLEN. [33,TruSeq3-PE.fa:2:30:10\n'
                                  '20,20,4:15,75]'),
                            metavar='')

        parser.add_argument('-e',
                            '--snpEff',
                            action='store',
                            type=str,
                            help=('Predict causal variant by SnpEff. Please check\n'
                                  'available databases in SnpEff.'),
                            metavar='')

        parser.add_argument('--mem',
                            action='store',
                            default='1G',
                            type=str,
                            help='Used memory size when bam sorted. [1G]',
                            metavar='')

        parser.add_argument('-q',
                            '--min-MQ',
                            action='store',
                            default=40,
                            type=int,
                            help='Minimum mapping quality in mpileup. [40]',
                            metavar='')

        parser.add_argument('-Q',
                            '--min-BQ',
                            action='store',
                            default=18,
                            type=int,
                            help='Minimum base quality in mpileup. [18]',
                            metavar='')

        parser.add_argument('-C',
                            '--adjust-MQ',
                            action='store',
                            default=50,
                            type=int,
                            help=('"adjust-MQ" in mpileup. Default parameter\n'
                                  'is suited for BWA. [50]'),
                            metavar='')

        # set version
        parser.add_argument('-v',
                            '--version',
                            action='version',
                            version='%(prog)s {}'.format(__version__))

        return parser

    def mutplot_options(self):
        parser = argparse.ArgumentParser(description='MutPlot version {}'.format(__version__),
                                         formatter_class=argparse.RawTextHelpFormatter)
        parser.usage = ('mutplot -v <VCF> -o <OUT_DIR> -n <INT> [-w <INT>] [-s <INT>]\n'
                        '               [-D <INT>] [-d <INT>] [-N <INT>] [-m <FLOAT>]\n'
                        '               [-S <INT>] [-e <DATABASE>] [--igv] [--indel]')

        # set options
        parser.add_argument('-v',
                            '--vcf',
                            action='store',
                            required=True,
                            type=str,
                            help='VCF which contains cultivar and mutant bulk.',
                            metavar='')

        parser.add_argument('-o',
                            '--out',
                            action='store',
                            required=True,
                            type=str,
                            help='Output directory. Specified name can exist.',
                            metavar='')

        parser.add_argument('-n',
                            '--N-bulk',
                            action='store',
                            required=True,
                            type=int,
                            help='Number of individuals in mutant bulk.',
                            metavar='')

        parser.add_argument('-w',
                            '--window',
                            action='store',
                            default=2000,
                            type=int,
                            help='Window size (kb). [2000]',
                            metavar='')

        parser.add_argument('-s',
                            '--step',
                            action='store',
                            default=100,
                            type=int,
                            help='Step size (kb). [100]',
                            metavar='')

        parser.add_argument('-D',
                            '--max-depth',
                            action='store',
                            default=250,
                            type=int,
                            help=('Maximum depth of variants which will be used.\n'
                                  'This cutoff will be applied in both of cultivar\n'
                                  'and bulk. [250]'),
                            metavar='')

        parser.add_argument('-d',
                            '--min-depth',
                            action='store',
                            default=8,
                            type=int,
                            help=('Minimum depth of variants which will be used.\n'
                                  'This cutoff will be applied in both of cultivar\n'
                                  'and bulk. [8]'),
                            metavar='')

        parser.add_argument('-N',
                            '--N-rep',
                            action='store',
                            default=10000,
                            type=int,
                            help=('Number of replicates for simulation to make \n'
                                  'null distribution. [10000]'),
                            metavar='')

        parser.add_argument('-m',
                            '--min-SNPindex',
                            action='store',
                            default=0.3,
                            type=float,
                            help='Cutoff of minimum SNP-index for clear results. [0.3]',
                            metavar='')

        parser.add_argument('-S',
                            '--strand-bias',
                            action='store',
                            default=7,
                            type=int,
                            help=('Filter spurious homo genotypes in cultivar using\n'
                                  'strand bias. If ADF (or ADR) is higher than this\n'
                                  'cutoff when ADR (or ADF) is 0, that SNP will be\n'
                                  'filtered out. If you want to supress this filtering,\n'
                                  'please set this cutoff to 0. [7]\n'),
                            metavar='')

        parser.add_argument('-e',
                            '--snpEff',
                            action='store',
                            type=str,
                            help=('Predict causal variant by SnpEff. Please check\n'
                                  'available databases in SnpEff.'),
                            metavar='')

        parser.add_argument('--igv',
                            action='store_true',
                            default=False,
                            help='Output IGV format file to check results on IGV.')

        parser.add_argument('--indel',
                            action='store_true',
                            default=False,
                            help='Plot SNP-index with INDEL.')

        parser.add_argument('--fig-width',
                            action='store',
                            default=7.5,
                            type=float,
                            help='Width allocated in chromosome figure. [7.5]',
                            metavar='')

        parser.add_argument('--fig-height',
                            action='store',
                            default=4.0,
                            type=float,
                            help='Height allocated in chromosome figure. [4.0]',
                            metavar='')

        parser.add_argument('--white-space',
                            action='store',
                            default=0.6,
                            type=float,
                            help=('White space between figures. (This option\n'
                                  'only affect vertical direction.) [0.6]'),
                            metavar='')

        # set version
        parser.add_argument('--version',
                            action='version',
                            version='%(prog)s {}'.format(__version__))

        return parser

    def check_max_threads(self, args):
        max_cpu = multi.cpu_count()
        print(time_stamp(),
              'max number of threads which you can use is {}.'.format(max_cpu),
              flush=True)
        if max_cpu <= args.threads:
            sys.stderr.write(('!!WARNING!! You can use up to {0} threads. '
                              'This program will use {0} threads.\n').format(max_cpu))
            sys.stderr.flush()
            args.threads = max_cpu
        elif args.threads < 1:
            args.threads = max_cpu
        return args

    def check_args(self, args):
        if os.path.isdir(args.out):
            sys.stderr.write(('  Output directory already exist.\n'
                              '  Please rename output directory or '
                                'remove existing directory\n\n'))
            sys.exit()

        N_fastq = 0

        for input_name in  args.cultivar:
            n_comma = input_name.count(',')
            if n_comma == 0:
                root, ext = os.path.splitext(input_name)
                if ext != '.bam':
                    sys.stderr.write(('  Please check "{}".\n'
                                      '  The extension of this file is not "bam".\n'
                                      '  If you wanted to specify fastq, please '
                                        'input them as paired-end reads which is separated '
                                        'by comma. e.g. -c fastq1,fastq2\n\n').format(input_name))
                    sys.exit()
            elif n_comma == 1:
                fastqs = input_name.split(',')
                for fastq in fastqs:
                    root, ext = os.path.splitext(fastq)
                    if ext == '.bam':
                        sys.stderr.write(('  Please check "{}".\n'
                                          '  The extension must not be "bam".\n'
                                          '  If you wanted to specify bam, please '
                                            'input them separately. e.g. -c bam1 '
                                            '-c bam2\n\n').format(input_name))
                        sys.exit()
                N_fastq += 1
            else:
                sys.stderr.write(('  Please check "{}".\n'
                                  '  You specified too much files, or '
                                    'your file name includes ",".\n\n').format(input_name))
                sys.exit()

        for input_name in  args.bulk:
            n_comma = input_name.count(',')
            if n_comma == 0:
                root, ext = os.path.splitext(input_name)
                if ext != '.bam':
                    sys.stderr.write(('  Please check "{}".\n'
                                      '  The extension of this file is not "bam".\n'
                                      '  If you wanted to specify fastq, please '
                                        'input them as paired-end reads which is separated '
                                        'by comma. e.g. -b fastq1,fastq2\n\n').format(input_name))
                    sys.exit()
            elif n_comma == 1:
                fastqs = input_name.split(',')
                for fastq in fastqs:
                    root, ext = os.path.splitext(fastq)
                    if ext == '.bam':
                        sys.stderr.write(('  Please check "{}".\n'
                                          '  The extension must not be "bam".\n'
                                          '  If you wanted to specify bam, please '
                                            'input them separately. e.g. -b bam1 '
                                            '-b bam2\n\n').format(input_name))
                        sys.exit()
                N_fastq += 1
            else:
                sys.stderr.write(('  Please check "{}".\n'
                                  '  You specified too much files, or '
                                    'your file name includes ",".\n\n').format(input_name))
                sys.exit()

        if N_fastq == 0 and args.trim:
            sys.stderr.write(('  You can specify "--trim" only when '
                                 'you input fastq.\n\n'))
            sys.exit()

        return N_fastq
